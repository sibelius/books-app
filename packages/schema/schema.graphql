"""Represents a Book"""
type Book implements Node {
  """The ID of an object"""
  id: ID!

  """MongoDB _id"""
  _id: String!

  """The book name. ex: O Alienista"""
  name: String

  """The book author. ex: Machado de Assis"""
  author: String

  """The book description."""
  description: String

  """The book release year. ex: 1882"""
  releaseYear: Int

  """The book total page. ex: 96"""
  pages: Int

  """The book price. ex: 20"""
  price: Float

  """The book banner url."""
  bannerUrl: String

  """The book banner ISBN. ex: 9780850515060"""
  ISBN: Int

  """The book language. ex: Portuguese"""
  language: String

  """The book average rating based on user reviews"""
  rating: Float

  """The soft delete status."""
  isActive: Boolean

  """The hard delete status."""
  removedAt: String

  """The date that the node was created."""
  createdAt: String

  """The date that the node was last updated."""
  updatedAt: String
}

"""A connection to a list of items."""
type BookConnection {
  """Number of items in this connection"""
  count: Int!

  """
  A count of the total number of objects in this connection, ignoring pagination.
    This allows a client to fetch the first five objects by passing "5" as the
    argument to "first", then fetch the total count so it could display "5 of 83",
    for example.
  """
  totalCount: Int!

  """Offset from start"""
  startCursorOffset: Int!

  """Offset till end"""
  endCursorOffset: Int!

  """Information to aid in pagination."""
  pageInfo: PageInfoExtended!

  """A list of edges."""
  edges: [BookEdge]!
}

"""An edge in a connection."""
type BookEdge {
  """The item at the end of the edge"""
  node: Book!

  """A cursor for use in pagination"""
  cursor: String!
}

"""Used to filter book"""
input BookFilters {
  OR: [BookFilters]
  AND: [BookFilters]

  """[{ direction: 'ASC', sort: 'CREATED_AT' }]"""
  orderBy: [BookOrdering!]

  """Filter by search string. Name, author or description."""
  search: String
}

"""Used to order books"""
input BookOrdering {
  sort: BookSortEnumType!
  direction: DirectionEnum!
}

enum BookSortEnumType {
  CREATED_AT
}

enum DirectionEnum {
  ASC
  DESC
}

input MeEditInput {
  """User name. ex: Jean"""
  name: String

  """Language of the user. ex: pt"""
  lang: String
  clientMutationId: String
}

type MeEditPayload {
  me: User

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

type Mutation {
  UserLogin(input: UserLoginInput!): UserLoginPayload
  UserRegistration(input: UserRegistrationInput!): UserRegistrationPayload
  MeEdit(input: MeEditInput!): MeEditPayload
  UserForgotPassword(input: UserForgotPasswordInput!): UserForgotPasswordPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfoExtended {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""The root of all... queries"""
type Query {
  """The ID of an object"""
  id: ID!

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID
  ): Node

  """Fetches objects given their IDs"""
  nodes(
    """The IDs of objects"""
    ids: [ID!]!
  ): [Node]!

  """Me is the logged User"""
  me: User
  status: Status

  """Connection to all books"""
  books(after: String, first: Int, before: String, last: Int, filters: BookFilters): BookConnection!

  """Connection to all reviews"""
  reviews(after: String, first: Int, before: String, last: Int, filters: ReviewFilters): ReviewConnection!
}

"""Represents a Review"""
type Review implements Node {
  """The ID of an object"""
  id: ID!

  """MongoDB _id"""
  _id: String!

  """The user who created this review."""
  user: User

  """The book rated on this review."""
  book: Book

  """The rating of the review. ex: 4.5"""
  rating: Float

  """The review description. ex: This book is awesome"""
  description: String

  """The soft delete status."""
  isActive: Boolean

  """The hard delete status."""
  removedAt: String

  """The date that the node was created."""
  createdAt: String

  """The date that the node was last updated."""
  updatedAt: String
}

"""A connection to a list of items."""
type ReviewConnection {
  """Number of items in this connection"""
  count: Int!

  """
  A count of the total number of objects in this connection, ignoring pagination.
    This allows a client to fetch the first five objects by passing "5" as the
    argument to "first", then fetch the total count so it could display "5 of 83",
    for example.
  """
  totalCount: Int!

  """Offset from start"""
  startCursorOffset: Int!

  """Offset till end"""
  endCursorOffset: Int!

  """Information to aid in pagination."""
  pageInfo: PageInfoExtended!

  """A list of edges."""
  edges: [ReviewEdge]!
}

"""An edge in a connection."""
type ReviewEdge {
  """The item at the end of the edge"""
  node: Review!

  """A cursor for use in pagination"""
  cursor: String!
}

"""Used to filter reviews"""
input ReviewFilters {
  OR: [ReviewFilters]
  AND: [ReviewFilters]

  """[{ direction: 'ASC', sort: 'CREATED_AT' }]"""
  orderBy: [ReviewOrdering!]

  """Filter by book."""
  book: ID

  """Filter by user."""
  user: ID
}

"""Used to order reviews"""
input ReviewOrdering {
  sort: ReviewSortEnumType!
  direction: DirectionEnum!
}

enum ReviewSortEnumType {
  """The rating of the review. ex: 4.5"""
  RATING
  CREATED_AT
}

"""Status"""
type Status {
  """Books app environment"""
  env: String

  """Server build version"""
  version: String

  """Server commit sha, GITHUB_SHA"""
  commitSha: String

  """Books app environment"""
  e: String

  """Server build version"""
  v: String

  """Server commit sha, GITHUB_SHA"""
  c: String
}

"""Represents an user"""
type User implements Node {
  """The ID of an object"""
  id: ID!

  """MongoDB _id"""
  _id: String!

  """User name resolver"""
  name: String

  """User surname resolver"""
  surname: String

  """User name resolver"""
  fullName: String

  """User email resolver"""
  email: String

  """User email resolver"""
  emailWasVerified: String
  lang: String

  """The soft delete status."""
  isActive: Boolean

  """The hard delete status."""
  removedAt: String

  """The date that the node was created."""
  createdAt: String

  """The date that the node was last updated."""
  updatedAt: String
}

input UserForgotPasswordInput {
  email: String!
  clientMutationId: String
}

type UserForgotPasswordPayload {
  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

input UserLoginInput {
  """User email to be used on login. ex: jean@gmail.com"""
  email: String!

  """User password."""
  password: String!
  clientMutationId: String
}

type UserLoginPayload {
  token: String

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

input UserRegistrationInput {
  """User name. ex: Jean"""
  name: String!

  """User email to be used on login. ex: jean@gmail.com"""
  email: String!

  """User password."""
  password: String!
  clientMutationId: String
}

type UserRegistrationPayload {
  token: String

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}
