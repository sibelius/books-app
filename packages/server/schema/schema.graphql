input MeEditInput {
  """User name. ex: Jean"""
  name: String

  """Language of the user. ex: pt"""
  lang: String
  clientMutationId: String
}

type MeEditPayload {
  me: User

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

type Mutation {
  UserLogin(input: UserLoginInput!): UserLoginPayload
  UserRegistration(input: UserRegistrationInput!): UserRegistrationPayload
  MeEdit(input: MeEditInput!): MeEditPayload
  UserForgotPassword(input: UserForgotPasswordInput!): UserForgotPasswordPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""The root of all... queries"""
type Query {
  """The ID of an object"""
  id: ID!

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID
  ): Node

  """Fetches objects given their IDs"""
  nodes(
    """The IDs of objects"""
    ids: [ID!]!
  ): [Node]!

  """Me is the logged User"""
  me: User
  status: Status
}

"""Status"""
type Status {
  """Books app environment"""
  env: String

  """Server build version"""
  version: String

  """Server commit sha, GITHUB_SHA"""
  commitSha: String

  """Books app environment"""
  e: String

  """Server build version"""
  v: String

  """Server commit sha, GITHUB_SHA"""
  c: String
}

"""Represents an user"""
type User implements Node {
  """The ID of an object"""
  id: ID!

  """MongoDB _id"""
  _id: String!

  """User name resolver"""
  name: String

  """User surname resolver"""
  surname: String

  """User name resolver"""
  fullName: String

  """User email resolver"""
  email: String

  """User email resolver"""
  emailWasVerified: String
  lang: String

  """The soft delete status."""
  isActive: Boolean

  """The hard delete status."""
  removedAt: String

  """The date that the node was created."""
  createdAt: String

  """The date that the node was last updated."""
  updatedAt: String
}

input UserForgotPasswordInput {
  email: String!
  clientMutationId: String
}

type UserForgotPasswordPayload {
  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

input UserLoginInput {
  """User email to be used on login. ex: jean@gmail.com"""
  email: String!

  """User password."""
  password: String!
  clientMutationId: String
}

type UserLoginPayload {
  token: String

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

input UserRegistrationInput {
  """User name. ex: Jean"""
  name: String!

  """User email to be used on login. ex: jean@gmail.com"""
  email: String!

  """User password."""
  password: String!
  clientMutationId: String
}

type UserRegistrationPayload {
  token: String

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}
