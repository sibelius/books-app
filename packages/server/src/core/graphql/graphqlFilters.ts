// inspired by: https://github.com/entria/graphql-mongo-helpers

import { ConnectionArguments } from 'graphql-relay';
import idx from 'idx';

import { JSObject } from '@booksapp/types';

import { GraphQLContext } from '../../types';
import { getObjectId } from '../../common/utils';

import { DIRECTION } from './enum/DirectionEnumType';

export const FILTER_CONDITION_TYPE = {
  MATCH_1_TO_1: 'MATCH_1_TO_1', // something that could be used on find() or $match
  CUSTOM_CONDITION: 'CUSTOM_CONDITION', // create a custom condition based on value
  AGGREGATE_PIPELINE: 'AGGREGATE_PIPELINE', // just an aggregate
};

export type BuiltConditionSet = {
  conditions: JSObject;
  pipeline: JSObject[];
};

type FilterFieldMappingMatch<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.MATCH_1_TO_1;
  key: string;
  format?: (value: ValueT, filter: JSObject, context: GraphQLContext, options?: JSObject<any>) => any;
};

type FilterFieldMappingPipeline<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE;
  pipeline: JSObject[] | ((value: ValueT) => Array<{}>);
};

type FilterFieldMappingCustomCondition<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.CUSTOM_CONDITION;
  format: (value: ValueT, filter: JSObject, context: GraphQLContext, options?: JSObject<any>) => any;
};

type FilterFieldMapping<ValueT> =
  | FilterFieldMappingMatch<ValueT>
  | FilterFieldMappingPipeline<ValueT>
  | FilterFieldMappingCustomCondition<ValueT>
  | boolean;

export type FilterMapping = { [key: string]: FilterFieldMapping<any> };

const validOperators = ['gt', 'gte', 'lt', 'lte', 'in', 'nin', 'ne', 'all'];

const arrayOperators = ['in', 'nin', 'all'];

const getFilterName = (filterName: string) => filterName.split('_')[0];

type Operators = 'gt' | 'gte' | 'lt' | 'lte' | 'in' | 'nin' | 'ne' | 'all' | '$and' | '$or';

const handleAndOr = (operator: Operators) => (
  context: GraphQLContext,
  condition: Array<{}>,
  mapping: { [key: string]: FilterFieldMapping<{}> },
) => {
  if (!Array.isArray(condition)) {
    throw new Error(`Invalid filter supplied to ${operator}.`);
  }

  return {
    condition: condition.map((andCondition) => buildConditionsObject(context, andCondition, mapping)),
    conditionName: operator,
  };
};

const handleAnd = handleAndOr('$and');
const handleOr = handleAndOr('$or');

const handleFieldOperator = (
  context: GraphQLContext,
  condition: JSObject,
  conditionName: string,
  fieldMapping: FilterFieldMapping<JSObject>,
  prev: JSObject,
  filters: JSObject,
  options = {},
) => {
  // { "myField_operator": "something" } becomes { "myField": { $operator: "something" } }
  // { "myField": "something" } remains the same
  const conditionNamePieces = conditionName.split('_');
  const operator = conditionNamePieces.length > 1 ? conditionNamePieces.pop() : '';
  // I don't think we support snake case for field names, should this be here?
  conditionName = conditionNamePieces.join('_');

  if (fieldMapping && fieldMapping.format && typeof fieldMapping.format === 'function') {
    condition = fieldMapping.format(condition, filters, context, options);
  }

  if (operator) {
    if (validOperators.indexOf(operator) === -1) {
      throw new Error(`"${operator}" is not a valid operator on field "${conditionName}".`);
    }

    if (arrayOperators.indexOf(operator) >= 0 && !Array.isArray(condition)) {
      throw new Error(`Field "${conditionName}" must have an array value.`);
    }

    condition = {
      [`$${operator}`]: condition,
    };
  }

  // handle $gte and $let fields merge
  if (conditionName in prev) {
    condition = {
      ...condition,
      ...prev[conditionName],
    };
  }

  return {
    condition,
    conditionName,
  };
};

function buildConditionsObject<ValueT>(
  context: GraphQLContext,
  conditions: JSObject,
  mapping: { [key: string]: FilterFieldMapping<ValueT> },
  filters?: JSObject,
  options = {},
) {
  return Object.keys(conditions).reduce((prev, currentKey) => {
    let condition = conditions[currentKey];
    let conditionName = currentKey;

    const fieldMapping = mapping[getFilterName(currentKey)];

    if (fieldMapping === false) {
      return prev;
    }

    if (
      fieldMapping &&
      fieldMapping.type !== FILTER_CONDITION_TYPE.MATCH_1_TO_1 &&
      fieldMapping.type !== FILTER_CONDITION_TYPE.CUSTOM_CONDITION
    ) {
      return prev;
    }

    if (conditionName === 'AND') {
      ({ condition, conditionName } = handleAnd(context, condition, mapping));
    } else if (conditionName === 'OR') {
      ({ condition, conditionName } = handleOr(context, condition, mapping));
    } else {
      if (fieldMapping && fieldMapping.type === FILTER_CONDITION_TYPE.CUSTOM_CONDITION) {
        if (fieldMapping.format && typeof fieldMapping.format === 'function') {
          const customCondition = fieldMapping.format(condition, filters, context, options);

          return {
            ...prev,
            ...customCondition,
          };
        }
      }
      ({ condition, conditionName } = handleFieldOperator(
        context,
        condition,
        conditionName,
        fieldMapping,
        prev,
        filters,
        options,
      ));
    }

    conditionName = fieldMapping && fieldMapping.key ? fieldMapping.key : conditionName;

    if (condition === undefined) {
      return prev;
    }

    return {
      ...prev,
      [conditionName]: condition,
    };
  }, {});
}

export function buildMongoConditionsFromFilters<ValueT>(
  context: GraphQLContext,
  filters: JSObject,
  mapping: { [key: string]: FilterFieldMapping<ValueT> } = {},
  options = {},
): BuiltConditionSet {
  if (!filters) {
    return { conditions: {}, pipeline: [] };
  }

  const keys = Object.keys(filters);

  // first check if there are any pipeline mapped fields
  // and if AND or OR are also passed, if that is the case, we must throw an error
  // because we cannot use OR/AND while also using pipeline.
  const hasPipelineFilter = keys.find(
    (item) => idx(mapping[item], (_) => _.type) === FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE,
  );
  if (hasPipelineFilter && (filters.AND || filters.OR)) {
    throw new Error(
      `Wrong filter usage, because filter "${hasPipelineFilter}" is a pipeline filter, which should disable AND and OR`,
    );
  }

  // separate filters by type
  const filtersKeysGrouped = Object.keys(filters).reduce(
    (prev, key) => {
      const filterName = getFilterName(key);

      const type = (mapping && idx(mapping[filterName], (_) => _.type)) || FILTER_CONDITION_TYPE.MATCH_1_TO_1;

      return {
        ...prev,
        [type]: {
          ...prev[type],
          [key]: filters[key],
        },
      };
    },
    {
      // start with sane defaults
      [FILTER_CONDITION_TYPE.MATCH_1_TO_1]: {},
      [FILTER_CONDITION_TYPE.CUSTOM_CONDITION]: {},
      [FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE]: {},
    },
  );

  // first build our conditions JSObject.
  const conditions = buildConditionsObject(
    context,
    { ...filtersKeysGrouped.MATCH_1_TO_1, ...filtersKeysGrouped.CUSTOM_CONDITION },
    mapping,
    filters,
    options,
  );

  // now build the pipeline, which is more straightforward
  const pipeline = Object.keys(filtersKeysGrouped.AGGREGATE_PIPELINE).reduce((prev, key) => {
    const mappedFilter = mapping[key];
    // should not really happen!
    if (!mappedFilter || mappedFilter.type !== FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE) {
      return prev;
    }

    const fieldPipeline = Array.isArray(mappedFilter.pipeline)
      ? mappedFilter.pipeline
      : mappedFilter.pipeline(filters[key], filters, context, options);

    return [...prev, ...(Array.isArray(fieldPipeline) ? fieldPipeline : [])];
  }, []);

  return {
    conditions,
    pipeline,
  };
}

type OrderByArg<SortT> = {
  sort: SortT;
  direction: typeof DIRECTION;
};

export function buildSortFromOrderByArg<Sort>(orderByArg: Array<OrderByArg<string>>): { [key: string]: 1 | -1 } {
  return orderByArg.reduce(
    (acc, item) => ({
      ...acc,
      [item.sort]: item.direction,
    }),
    {},
  );
}

// @TODO - treat when there's more than one $match
export function orderAggregatePipeline(pipeline: JSObject[]): JSObject[] {
  if (!pipeline.length) {
    return [];
  }

  // @TODO - support $geoWithin as well
  const geoNearStage = pipeline.find((stage) => !!stage.$geoNear);

  let orderedPipeline: JSObject[] = [];
  let conditions = {};
  for (const stage of pipeline) {
    // IMPORTANT: if the stage is geoNear put it as first of the pipeline
    if (!!stage.$geoNear) {
      orderedPipeline = [stage, ...orderedPipeline];
      continue;
    }

    // store the first appearing $match on conditions
    // discard it from the ordered pipeline
    if (!!stage.$match) {
      if (geoNearStage && Object.keys(conditions).length === 0) {
        conditions = { ...conditions, ...stage.$match };
        continue;
      }
    }

    orderedPipeline = [...orderedPipeline, stage];
  }

  // add conditions on query of $geoNear
  if (geoNearStage) {
    geoNearStage.$geoNear.query = { ...(geoNearStage.$geoNear.query || {}), ...conditions };
  }

  return orderedPipeline;
}

type ArgsWithFilter = {
  filters: { [key: string]: string };
} & { [key: string]: string } & ConnectionArguments;
export const withFilter = (args: ArgsWithFilter, filters: JSObject) => {
  return {
    ...args,
    filters: {
      ...(args.filters || {}),
      ...filters,
    },
  };
};

export const getFieldsOneToOne = (fields: string[], JSObjectId = false) => {
  const formatParam = JSObjectId ? { format: (value?: string) => value && getObjectId(value) } : {};

  return fields.reduce(
    (acc, field) => ({
      ...acc,
      [field]: {
        type: FILTER_CONDITION_TYPE.MATCH_1_TO_1,
        key: field,
        ...formatParam,
      },
    }),
    {},
  );
};
