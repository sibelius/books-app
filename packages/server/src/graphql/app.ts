/* eslint-disable no-console */
import 'isomorphic-fetch';

import { koaPlayground } from 'graphql-playground-middleware';
import { print } from 'graphql/language';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from '@koa/cors';
import graphqlHttp, { OptionsData } from 'koa-graphql';
import koaLogger from 'koa-logger';
import multer from 'koa-multer';
import Router from '@koa/router';
import idx from 'idx';
import i18next from 'i18next';
import koaI18next from 'koa-i18next';
import NoIntrospection from 'graphql-disable-introspection';

import * as graphqlLoaders from '../loader';

import { JWT_KEY, DEBUG_GRAPHQL } from '../common/config';

import { GraphQLContext, KoaContextExt } from '../types';

import UserModel, { IUser } from '../modules/user/UserModel';

import { t } from '../locales/helpers';

import appPlatformMiddleware from '../modules/auth/appPlatformMiddleware';
import authMiddleware from '../modules/auth/authMiddleware';

import { getDataloaders, hasIntrospectionQuery } from './helper';
import { schema } from './schema';
import { i18nMiddleware } from './i18n';

const app = new Koa<any, KoaContextExt>();
// https://github.com/koajs/koa/blob/422e539e8989e65ba43ecc39ddbaa3c4f755d465/docs/api/request.md#requestip
if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}
app.keys = [JWT_KEY];

const loaders = graphqlLoaders;

const router = new Router<any, KoaContextExt>();
const storage = multer.memoryStorage();

// https://github.com/expressjs/multer#limits
const limits = {
  // Increasing max upload size to 30 mb, since busboy default is only 1 mb
  fieldSize: 30 * 1024 * 1024,
};

app.use(bodyParser());

// https://github.com/koajs/koa/wiki/Error-Handling
// When using `koa-graphql` this is not really called, but let's keep in in here
// for future reference
// see https://github.com/chentsulin/koa-graphql/issues/85
app.use(async (ctx, next) => {
  try {
    await next();
  } catch (err) {
    console.log('koa error:', err);
    ctx.status = err.status || 500;
    ctx.app.emit('error', err, ctx);
  }
});

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.get('/healthz', async (ctx) => {
  try {
    await UserModel.find().lean<IUser>();
    ctx.body = 'Database working';
    ctx.status = 200;
  } catch (err) {
    ctx.body = err.toString();
    ctx.status = 500;
  }
});

router.all('/graphql', multer({ storage, limits }).any());

if (process.env.NODE_ENV !== 'production') {
  router.all(
    '/playground',
    koaPlayground({
      endpoint: '/graphql',
    }),
  );
}

// Middleware to get dataloaders
app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang', // detecting language in cookie
    order: ['cookie', 'header'],
    next: true,
  }),
);

app.use(i18nMiddleware);

// avoid requests from unknown clients
if (process.env.NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

// Middleware to get result from authorization token
app.use(authMiddleware);

router.all(
  '/graphql',
  convert(
    graphqlHttp(
      async (request, ctx, koaContext): Promise<OptionsData> => {
        const { dataloaders, user } = koaContext;
        const { authorization, appversion, appbuild, appplatform, timezone = 'America/Sao_Paulo' } = request.header;

        if (process.env.NODE_ENV !== 'test') {
          console.info('Handling request', {
            appversion,
            appbuild,
            appplatform,
            timezone,
            unauthorized: !user,
            authorization,
            userId: idx(user, (_) => _._id),
          });
        }

        return {
          graphiql: process.env.NODE_ENV === 'development',
          schema,
          rootValue: { request: ctx.req },
          validationRules: process.env.NODE_ENV === 'production' ? [NoIntrospection] : [],
          context: {
            user,
            dataloaders,
            appplatform,
            koaContext,
            timezone,
            t,
          } as GraphQLContext,
          extensions: ({ document, variables, result }) => {
            if (process.env.NODE_ENV === 'development' && DEBUG_GRAPHQL) {
              if (document && !hasIntrospectionQuery(document)) {
                console.log(print(document));

                console.log(variables);

                console.log(JSON.stringify(result, null, 2));
              }
            }
            // the type of extensions does not accept null, but the code does
            return null as any;
          },
          formatError: (error: any) => {
            if (error.name && error.name === 'BadRequestError') {
              ctx.status = 400;
              ctx.body = 'Bad Request';
              return {
                message: 'Bad Request',
              };
            }

            if (error.path || error.name !== 'GraphQLError') {
              console.error(error);
            } else {
              console.log(`GraphQLWrongQuery: ${error.message}`);
            }

            console.error('GraphQL Error', { error });

            if (process.env.NODE_ENV !== 'production') {
              return {
                message: error.message,
                locations: error.locations,
                stack: error.stack,
              };
            } else {
              ctx.status = 400;
              ctx.body = 'Bad Request';
              return {
                message: 'Bad Request',
              };
            }
          },
        };
      },
    ),
  ),
);

app.use(router.routes()).use(router.allowedMethods());

export default app;
