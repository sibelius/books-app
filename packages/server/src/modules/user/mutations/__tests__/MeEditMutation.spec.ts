import { graphql } from 'graphql';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { schema } from '../../../../graphql/schema';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
  gql,
} from '../../../../../test/helpers';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('MeEditMutation', () => {
  it('should edit context user', async () => {
    const user = await createUser();

    const mutation = gql`
      mutation M($input: MeEditInput!) {
        MeEdit(input: $input) {
          me {
            id
            email
            lang
          }
          error
        }
      }
    `;

    const variables = {
      input: {
        name: 'Awesome Name',
        lang: 'pt',
      },
    };
    const rootValue = {};
    const context = await getContext({ user });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.MeEdit.error).toBe(null);
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });
});
