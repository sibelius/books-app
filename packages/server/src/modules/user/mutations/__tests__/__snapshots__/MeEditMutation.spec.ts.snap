// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeEditMutation should edit context user 1`] = `
Object {
  "data": Object {
    "MeEdit": Object {
      "error": "EMPTY",
      "me": Object {
        "email": "user1@example.com",
        "id": "FROZEN-ID",
        "lang": "pt",
      },
    },
  },
}
`;
